quarkchain/evm/state.py

```diff
@@ -92,49 +92,120 @@ class TokenBalancePair(rlp.Serializable):
+
+    def commit(self):
+        self._committed = True
+        # No-op if not using trie and token size is small
+        if not self.token_trie and self._non_zero_entries <= TOKEN_TRIE_THRESHOLD:
+            return
+        # Use trie if size exceeds threshold
+        if not self.token_trie:
+            self.token_trie = SecureTrie(Trie(self._db))
+
+        keys_to_del = []
+        for token_id, bal in self._balances.items():
+            k = utils.encode_int32(token_id)
+            if bal:
+                self.token_trie.update(k, rlp.encode(bal))
+            else:
+                self.token_trie.delete(k)
+                keys_to_del.append(token_id)
+        for token_id in keys_to_del:
+            del self._balances[token_id]
+
+        self.token_root = self.token_trie.root_hash
 
     def serialize(self):
-        if len(self.balances) == 0:
+        if not self._committed:
+            self.commit()
+
+        # Fast path, since already committed
+        if self.token_trie:
+            return b"\x01" + self.token_root
+
+        # Serialize in-memory balance representation as an array
+        if len(self._balances) == 0:
             return b""
-        retv = self.enum
-        if self.enum == b"\x00":
-            l = []
-            for k, v in self.balances.items():
-                # Don't serialize 0 balance
-                if v == 0:
-                    continue
-                l.append(TokenBalancePair(k, v))
-            # sort by token id to make token balances serialization deterministic
-            l.sort(key=lambda b: b.token_id)
-            retv = retv + rlp.encode(l)
-        elif self.enum == b"\x01":
-            raise Exception("Token balance trie is not yet implemented")
-        else:
-            raise Exception("Unknown enum byte in token_balances")
-        return retv
+        # Don't serialize 0 balance
+        ls = [TokenBalancePair(k, v) for k, v in self._balances.items() if v > 0]
+        # Sort by token id to make token balances serialization deterministic
+        ls.sort(key=lambda b: b.token_id)
+        return b"\x00" + rlp.encode(ls)
 
     def balance(self, token_id):
-        return self.balances.get(token_id, 0)
+        if token_id in self._balances:
+            return self._balances[token_id]
+        if self.token_trie:
+            v = self.token_trie.get(utils.encode_int32(token_id))
+            ret = utils.big_endian_to_int(rlp.decode(v)) if v else 0
+            self._balances[token_id] = ret
+            return ret
+        return 0
+
+    def set_balance(self, journal, token_id, value):
+        preval = self.balance(token_id)
+        self._balances[token_id] = value
+        journal.append(lambda: self._balances.__setitem__(token_id, preval))
+        if self._committed:
+            self._committed = False
+            journal.append(lambda: setattr(self, "_committed", True))
+
+    def is_blank(self):
+        return not self.token_trie and self._balances == {}
```

Comment:
We probably need _non_zero_entries == 0 to make sure zero balance, a further comment may be needed to explain if token_trie is not NULL_TRIE_HASH.

Reply:
done